firmware/isp.c

```diff
@@ -124,21 +125,31 @@ void ispConnect() {
 
 void ispDisconnect() {
 
+	cli();
 	/* set all ISP pins inputs */
 	ISP_DDR &= ~((1 << ISP_RST) | (1 << ISP_SCK) | (1 << ISP_MOSI));
 	/* switch pullups off */
 	ISP_OUT &= ~((1 << ISP_RST) | (1 << ISP_SCK) | (1 << ISP_MOSI));
+	sei();
 
 	/* disable hardware SPI */
 	spiHWdisable();
 }
 
 void ispSafeResetPulse(void)
 {
+	/* 
```

Comment:
Actually the AVR can do atomic bit sets and clears.
So It is more complex than  just read-modify-write issue.

The AVR has atomic bit set/bit clear instructions. They can only set or clear a single bit.
Many other processors (pic32, arm) use bit set/clr registers so you can operate on multiple bits
avr-gcc has a giant kludge in it that when optimization is enabled, the compiler will generate a single atomic bit set/bit clear instruction, instead of multiple in/read, or, out/write  instructions whenever all the parameters of the |= or &= are known at compile time and conditions allow it.
So if you do something like 

```ISP_OUT |= (1<<ISP_RST);```

since the memory address of the underlying pointer and the bit mask are composed of constants and known at compile time and the pointer is in an address that lands within the bit set/clr address range, the optimizer will reduce that to a single  sbr (bit set) instruction.
When a single instruction is used to set/clear a bit in a register, the operation is atomic.
(like I said it is a massive kludge)
Same is true on the clear side with code like:

```ISP_OUT &= ~(1<<ISP_RST);```

The catch is that this kludge optimization only works on single bits. So that is why I added cli()/sei() around the multi-bit operations. Those are interruptible as they are multiple instructions.
If you look in the generated code in ipSafeResetPulse() you will see the sbi/cbi instructions for handling of the ISP_RST signal.

I spent 9 months trying to get Dave Mellis to fix the arduino code in digitalWrite() to correct an atomicity issue. He didn't understand this concept and to this day I still don't think he does. He only reluctantly fixed the digitalWrite() code because with the update in the digitalWrite() code, the servo library worked reliably.
He never could understand how an issue in foreground code could break code running at ISR level.

One really particularly nasty part of all this is that for pats like the mega 640/1280/2560
Not all the ports are in the address range to support the sbi/cbi instructiond. So on those parts you have be VERY careful since most ports that will use sbi/cbi but some will use sbi/cbi.
My view bit set/clear instructions are not useful outside of assembler programming and  set/clear registers is a much better way to go since not only can it support multiple bits but it can work in higher level languages without having to do anything special to the compiler.


Although I just noticed I missed a location in ispConnect() so I'll need to go fix that.